ENH: Annotate the arithmetic operations of `ndarray` and `generic`

BvB93 opened the pull request:
This pull requests adds annotations for `ndarray` and `generic` arithmetic operations:

* `__add__`
* `__sub__`
* `__mul__`
* `__truediv__`
* `__floordiv__`
* `__pow__`

A few notes about the PR:

While adding annotations to `timedelta64` I encountered numerous issue due to its incompatibility with its `signedinteger` baseclass: 
* No support for `__index__` 
* No support for `__pow__`
* Does not support operations with complex numbers
* Most operations do not support floats

For this reason its baseclass has been changed from `signedinteger` to `generic`. While not strictly true during runtime, the problems with `timedelta64` inheriting from `signedinteger` have been noted before (https://github.com/numpy/numpy/issues/10685) and this change would be inline with the fix proposed in abovementioned issue.

Secondly, one obstacle that hasn't been dealt with is how the handle `generic`s precision, 
_e.g._ how to annotate `int64(...) + float32(...) -> float64(...)` without adding massive amounts of `overload`s. 
I might have an idea about to do deal with this issue, but I'd prefer to save it for a future PR due to some necessary compromises.

In any case, for the time being the return types are generally limited to types with ambiguous precision (`signedinteger`, `floating`, `complexfloating`, _etc..._)

``` python
import numpy as np

i8 = np.int64(...)
f4 = np.float32(...)
f4 + i8  # note: Revealed type is 'numpy.floating'
```

Finally, in order the deal with the repetitive and `@overload`-prone function signatures, this pull request makes extensive use of [callback protocols](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols). Callback protocols are effectively more fancy version of `typing.Callable` and, therefore , can be assigned and reused.

Commit:
ENH: Added annotations for arithmetic-based magic methods

Commit:
TST: Added arithmetic tests

Commit:
TST: Moved a number of tests to `arithmetic.py`

Commit:
ENH: Ensure that objects annotated as `number` support arithmetic operations

Commit:
MAINT: Arithmetic operations on 0d arrays return scalars

Commit:
MAINT: Clarify the type of generics returned by `ufunc.__call__`

Commit:
TST: Added more arithmetic tests

Commit:
MAINT: Use `_CharLike` when both `str` and `bytes` are accepted

Commit:
MAINT: Change the `timedelta64` baseclass to `generic`

Commit:
MAINT: Add aliases for common scalar unions

Commit:
MAINT: Update the defition of `_NumberLike`

Commit:
MAINT: Replace `_NumberLike` with `_ComplexLike` in the `complexfloating` annotations

Commit:
MAINT: Move the callback protocols to their own module

Commit:
MAINT: Make `typing._callback` available at runtime

eric-wieser commented on numpy/typing/_callback.py:
```diff
@@ -0,0 +1,132 @@
+"""A module with various callback protocols."""
```
Do you mean "callable"? It looks to me like the goal here is to provide common overloaded function / method signatures, if that helps you clarify the comment.

BvB93 replied:
The [mypy docs](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols) refer to these type of protocols as callback protocols, 
presumably because (them being classes) protocols are always callable objects.

I expanded the module-level docstring a bit in https://github.com/numpy/numpy/pull/17273/commits/8608127f59518411ab0e5e8e0c5059ac839b0c0a in order to provide further clarity on the subject.

eric-wieser replied:
I disagree - the mypy docs are showing how this type of protocol can be _used_ for callbacks. But in the numpy stubs here, these _aren't_ callabacks. A callback is a callable that is a function argument, which these are not.

> presumably because (them being classes) protocols are always callable objects.

You're conflating the class itself, (which is callable) vs class instances (which are only callable if they define `__call__`).

BvB93 replied:
> A callback is a callable that is a function argument, which these are not.

I wasn't aware of this distinction; I see your point.
Fixed in https://github.com/numpy/numpy/pull/17273/commits/58c81b38b293370edbeee4552d12951a5792ca1e.

Commit:
DOC: Provide further clarification about callback protocols

Commit:
MAINT: Replace `_callback` with `_callable`

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r485821346

The use of `__call__`-defining protocols is not limited to callbacks. The module name name & docstring now reflects this.

Commit:
MAINT: Removed `__add__` from `str_` and `bytes_`

Most `np.bytes_` / `np.str_` methods return their builtin `bytes` / `str` counterpart.
This includes addition.

eric-wieser commented on numpy/typing/_callable.py:
```diff
@@ -0,0 +1,140 @@
+    class _BoolTrueDiv(Protocol):
+        @overload
+        def __call__(self, __other: _BoolLike) -> float64: ...
+        @overload  # platform dependent
+        def __call__(self, __other: int) -> Union[float32, float64]: ...
```
This isn't true, the result is always float64 as `np.result_type(np.int32, np.float32)` is `float64`

eric-wieser replied:
I think (but haven't checked) the truth is more like
```suggestion
        @overload
        def __call__(self, __other: Union[int8]) -> float16: ...
        @overload
        def __call__(self, __other: Union[int16]) -> float32: ...
        @overload
        def __call__(self, __other: Union[int, int32, int64]) -> float64: ...
```

BvB93 replied:
I just did some double a quick test locally and it seems the return type `np.bool_` by `np.integer` division is always `float64`:
```python
In [1]: import numpy as np

In [2]: int_types = [
   ...:     int,
   ...:     np.int8,
   ...:     np.int16,
   ...:     np.int32,
   ...:     np.int64,
   ...:     np.longlong,
   ...:     np.uint8,
   ...:     np.uint16,
   ...:     np.uint32,
   ...:     np.uint64,
   ...:     np.ulonglong
   ...: ]

In [3]: true = np.bool_(True)

In [4]: {type(true / typ(1)) for typ in int_types}
Out[4]: {numpy.float64}

In [5]: {type(typ(1) / true) for typ in int_types}                                                                                                                                              
Out[5]: {numpy.float64}
```

BvB93 replied:
Fixed in https://github.com/numpy/numpy/pull/17273/commits/06c745f95439b046714aab0c5a3ee44b56b97bcc.

eric-wieser commented on numpy/typing/_callable.py:
```diff
@@ -0,0 +1,140 @@
+    _BoolTrueDiv = Any
+    _TD64Div = Any
+    _IntTrueDiv = Any
+    _UnsignedIntArithmetic = Any
+    _SignedIntArithmetic = Any
+    _FloatArithmetic = Any
+    _ComplexArithmetic = Any
+    _NumberArithmetic = Any
```
I think these names would be better  with an `Op` or `Method` suffix

BvB93 replied:
Nice and short; I like it.

BvB93 replied:
Fixed in https://github.com/numpy/numpy/pull/17273/commits/9a9ce45adc99cfec5e9342508d3adf83cab42e78.

Commit:
MAINT: Fix the return type of boolean division

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r486271220

Dividing a `np.bool_` by an integer (or vice versa) always returns `float64`

Commit:
MAINT: Renamed all `_<x>Arithmetic` protocols to `_<x>Op

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r486272745

Commit:
TST: Add tests for boolean division

BvB93 commented:
Rebased to get rid of a merge conflict.

@eric-wieser could you mark the PR as approved if you've got no further comments?

Commit:
ENH: Make `np.number` generic w.r.t. its precision

Commit:
ENH,WIP: Add a mypy plugin for casting `np.number` instances to appropiate subclasses

Commit:
Revert "ENH,WIP: Add a mypy plugin for casting `np.number` instances to appropiate subclasses"

This reverts commit c526fb619d20902bfd77709c8983c7a7d5477c95.

Commit:
Revert "ENH: Make `np.number` generic w.r.t. its precision"

This reverts commit dbf20183cf7ff71e379cd1a165d07e1a1d643135.

BvB93 commented:
Whoops, wrong branch. Please ignore https://github.com/numpy/numpy/pull/17273/commits/dbf20183cf7ff71e379cd1a165d07e1a1d643135 and https://github.com/numpy/numpy/pull/17273/commits/c526fb619d20902bfd77709c8983c7a7d5477c95.

charris commented:
Using `git rebase -i HEAD~n`, where `n` is the number or commits you want to clean up, followed by a force push is usually cleaner than reverting. You can merge/delete commits that way.

BvB93 commented:
> You can merge/delete commits that way.

Good to know, I'll keep it in mind for the future.

eric-wieser commented on numpy/__init__.pyi:
```diff
@@ -541,6 +552,11 @@ class ndarray(_ArrayOrScalarCommon, Iterable, Sized, Container):
 
 _CharLike = Union[str, bytes]
+_BoolLike = Union[bool, bool_]
+_IntLike = Union[int, integer]
+_FloatLike = Union[int, integer, float, floating]
+_ComplexLike = Union[int, float, complex, number]
```
I'm wary of using `number` in this last one. I'd be inclined to write this
```suggestion
_IntLike = Union[int, integer]
_FloatLike = Union[_IntLike, float, floating]
_ComplexLike = Union[_FloatLike, complex, complexfloating]
```
While all numbers may be coercible to complex today, third-party types like `quaternion` will not be.

BvB93 replied:
Fair point, better to be on the safe side here. Implemented in https://github.com/numpy/numpy/pull/17273/commits/0e4a26771e787cc85232965cdfea6592bcef1526.

As a side note: there is a limit to allowed levels of `typing.Union` nesting, which is why I chose not to reuse `_IntLike` and `FloatLike` in the definition of `_ComplexLike`, despite the minor increase in code duplication.

eric-wieser replied:
I assume that limit is a mypy issue? That sucks. Mind adding a comment to that effect in the code?

BvB93 replied:
While I recall seeing this issue in the past I can't seem to reproduce it anymore; I suspect it has been fixed in one of the more recent mypy releases?

In any case, https://github.com/numpy/numpy/pull/17273/commits/4ab85b4e7da0b3e48d519cdc173a57234868e9ce introduced the change you initially suggested. 
If, at some point, any nesting related issues appear after all then we can always get rid of the nesting.

eric-wieser commented on numpy/typing/tests/data/reveal/arithmetic.py:
```diff
@@ -0,0 +1,256 @@
+reveal_type(i + f4)  # E: floating
+reveal_type(AR + f4)  # E: Union[numpy.ndarray, numpy.generic]
+
+# Int
+
+reveal_type(i8 + i8)  # E: signedinteger
+reveal_type(i8 + u8)  # E: Union[numpy.signedinteger, numpy.floating]
```
Note that the output I'd expect is
```python
reveal_type(i8 + u8)  # E: numpy.int16
```
Fine to dismiss that as future work, but a comment next to `# Int` or at the top of the file saying that many of these patterns are overly vague would be nice.

BvB93 replied:
The `floating` case is specifically for `uint64 + signedinteger -> float64`, though I suppose this does mean that `floating` can be replaced with `float64`.

https://github.com/numpy/numpy/pull/17273/commits/9e63ecf75d6055c3f8b424e7badb28773455b082 contains a few additional notes on the subject in addition to the `floating` / `float64` swap.

BvB93 replied:
And yes, further refining of the return types (_e.g._ `int64 + float32 -> float64`) is planned for a future PR.
Just have to think a bit first about be the best way to implement this.

Commit:
MAINT: Narow the definition of `_ComplexLike`

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r490440238

Commit:
MAINT: Refined the return type of `unint + int` ops

`unsignedinteger + signedinteger` generally returns a `signedinteger` subclass.
The exception to this is `uint64 + signedinteger`, which returns `float64`.

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r490442023

Commit:
MAINT: Use `_IntLike` and `_FloatLike` in the definition of `_ComplexLike`

charris commented:
@eric-wieser Are you OK with this?

BvB93 commented:
I think we've discussed the most important aspects. 
The main thing still on the to-do list is dealing infering the return types' precision, which is intended for a future PR
(_e.g._ `int64 + float32 -> float64` instead of `... -> floating`).

charris commented:
> I think we've discussed the most important aspects

OK, lets put this in, don't want to let too much work stack up :) Thanks @BvB93 .
